### PR TITLE
Allow Pulsar default WaitStrategy to honour startup timeout

### PR DESCRIPTION
**Pull request by nahguam**
## Motivation

Currently, the PulsarContainer WaitStrategy is configured after the user
has started the container. This means that any modifications that the
user makes to the WaitStrategy or startupTimeout are discarded.

## Modifications

This change honours the user's modifications to the WaitStrategy or
startupTimeout.

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

**Review by eddumelendez — CHANGES_REQUESTED** [review]
thanks for the PR @nahguam ! can you add a test, please? in order to make sure the startupTimeout is honored? Maybe just setting to 0 and expect an exception with the right message will be ok.

**Review by VMarisevs — APPROVED** [review]
I am having the same issue with hard override of wait strategy 
https://github.com/testcontainers/testcontainers-java/issues/5830

File: modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java [hunk 1]
```diff
@@ -96,23 +102,29 @@ protected void setupCommandAndEnv() {
...
         }
         if (functionsWorkerEnabled) {
-            waitStrategies.add(Wait.forLogMessage(".*Function worker service started.*", 1));
+            waitAllStrategy.withStrategy(Wait.forLogMessage(".*Function worker service started.*", 1));
         }
-        final WaitAllStrategy compoundedWaitStrategy = new WaitAllStrategy();
-        waitStrategies.forEach(compoundedWaitStrategy::withStrategy);
-        waitingFor(compoundedWaitStrategy);
+
+        if (startupTimeout != null) {
```
Comment (VMarisevs) [inline modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:123]: tbh I am not sure if this is a good strategy.
I would rely more on the health end-point 200 response from `/admin/v2/namespaces/public`

sort of:
```
 Wait.forHttp("/admin/v2/namespaces/public")
                .forStatusCode(200)
                .forPort(PulsarContainer.BROKER_HTTP_PORT)
                .withStartupTimeout(Duration.ofMinutes(3)
```

File: modules/pulsar/src/test/java/org/testcontainers/containers/PulsarContainerTest.java [hunk 2]
```diff
@@ -144,6 +145,14 @@ public void testClusterFullyInitialized() throws Exception {
         }
     }
 
+    @Test
+    public void testStartupTimeoutIsHonored() {
+        try (PulsarContainer pulsar = new PulsarContainer(PULSAR_IMAGE).withStartupTimeout(Duration.ZERO)) {
```
Comment (kiview) [inline modules/pulsar/src/test/java/org/testcontainers/containers/PulsarContainerTest.java:150]: Clever test, I like it 👍

File: modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java (hunk 1 above)
Comment (kiview) [reply modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:123]: Applying the timeout on the `WaitAllStrategy` is definitely the way to go here. `WaitAllStrategy` already provided different timeout strategies, with `WITH_OUTER_TIMEOUT` being the default.

The implementation of `WaitAllStrategy` is slightly failsafe, meaning `WITH_OUTER_TIMEOUT` will work as expected if you call `withStartupTimeout` *after* adding all strategies, but also working fine when adding strategies *after* `withStartupTimeout` has been applied:
https://github.com/testcontainers/testcontainers-java/blob/de1324ed2800eff4da326d0c23d281399d006bc0/core/src/main/java/org/testcontainers/containers/wait/strategy/WaitAllStrategy.java#L70-L77

File: modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java [hunk 3]
```diff
@@ -78,6 +78,12 @@ public PulsarContainer withTransactions() {
         return this;
     }
 
+    @Override
+    public PulsarContainer withStartupTimeout(Duration startupTimeout) {
+        this.startupTimeout = startupTimeout;
+        return super.withStartupTimeout(startupTimeout);
+    }
```
Comment (kiview) [inline modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:85]: We can omit this method and already set the `WaitAllStrategy` in the constructor. The `withStartupTimeout` method of `GenericContainer` will work as expected.

**Review by kiview — CHANGES_REQUESTED** [review]
Thanks a lot for the PR @nahguam. I think if we change the implementation to set the `WaitAllStrategy` in the constructor, it will cleanup the code (we don't need a custom `withStartupTimeout` implementation) and make the container behave in general more as expected with regards to custom configuration by the user.

File: modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java (hunk 1 above)
Comment (nahguam) [reply modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java]: Thanks for the great feedback and suggestions. PTAL

**Review by eddumelendez — APPROVED** [review]

**Review by kiview — APPROVED** [review]
Thanks for applying all the recommendations @nahguam, was great working with you 👍

**Pull request merged**